src/realm/sync/noinst/client_impl_base.hpp

```diff
@@ -1511,6 +1562,36 @@ inline bool ClientImpl::Session::check_received_sync_progress(const SyncProgress
     return check_received_sync_progress(progress, error_code);
 }
 
+inline Trigger::Trigger(network::Service* service, std::function<void()>&& handler)
+    : m_service(service)
+    , m_handler(std::move(handler))
+    , m_triggered(false)
+{
+}
+
+inline void Trigger::trigger()
+{
+    REALM_ASSERT(m_service);
+
+    util::LockGuard lock{m_mutex};
+    if (m_triggered) {
+        return;
+    }
+    m_triggered = !m_triggered;
+
+    auto handler = [self_weak = weak_from_this()] {
+        // Do not execute the handler if the Trigger does not exist anymore.
+        if (auto self = self_weak.lock()) {
```

Comment:
I like using the weak reference - would a bind_ptr be a better implementation so we don't need to worry about shared_ptr's for Trigger?

Reply:
I looked a bit into bind_ptr. I didn't fully get how it works, but I'll try to see if I can make it work with one.

Reply:
basically you just inherit from AtomicRefCountBase<> and then call util::make_bind<>() like you would std::make_shared() and it all just works. there aren't weak_ptr's, but if you followed my suggestion about moving the ptr to a separate struct then you can fix this by just having an enum for m_triggered with a destroyed state rather than a bool.

Reply:
wouldn't the bind_ptr just keep the trigger alive then? we only need to know when it was destructed. At the same time, I guess there is no harm if we call the handler one more time before the trigger is destructed.